alf/algorithms/icm_algorithm.py

```diff
@@ -0,0 +1,163 @@
+        else:
+            self._num_actions = action_spec.shape[-1]
+
+        self._action_spec = action_spec
+        self._reward_scale = 1.0
+
+        feature_dim = tf.nest.flatten(feature_spec)[0].shape[-1]
+
+        self._encoding_net = encoding_net
+
+        if forward_net is None:
+            encoded_action_spec = tensor_spec.TensorSpec((self._num_actions, ),
+                                                         dtype=tf.float32)
+            forward_net = EncodingNetwork(
+                name="forward_net",
+                input_tensor_spec=[feature_spec, encoded_action_spec],
+                fc_layer_params=(hidden_size, ),
+                last_layer_size=feature_dim)
+
+        self._forward_net = forward_net
+
+        if inverse_net is None:
+            inverse_net = EncodingNetwork(
+                name="inverse_net",
+                input_tensor_spec=[feature_spec, feature_spec],
+                fc_layer_params=(hidden_size, ),
+                last_layer_size=self._num_actions,
+                last_kernel_initializer=tf.initializers.Zeros())
+
+        self._inverse_net = inverse_net
+
+        self._reward_normalizer = ScalarAdaptiveNormalizer(
+            speed=reward_adapt_speed)
+
+    def _encode_action(self, action):
+        if tensor_spec.is_discrete(self._action_spec):
+            return tf.one_hot(indices=action, depth=self._num_actions)
+        else:
+            return action
+
+    def _variables(self):
+        vars = self._forward_net.variables + self._inverse_net.variables
+        if self._encoding_net:
+            vars += self._encoding_net.variables
+        return vars
+
+    def train_step(self, inputs, state):
+        """
+        Args:
+            inputs (tuple): observation and previous action
+        Returns:
+            TrainStep:
+                outputs: intrisic reward
+                state:
+                info:
+        """
+        feature, prev_action = inputs
+        if self._encoding_net is not None:
+            feature, _ = self._encoding_net(feature)
+        prev_feature = state
+        prev_action = self._encode_action(prev_action)
+
+        forward_pred, _ = self._forward_net(
+            inputs=[tf.stop_gradient(prev_feature), prev_action])
+        forward_loss = 0.5 * tf.reduce_mean(
+            tf.square(tf.stop_gradient(feature) - forward_pred), axis=-1)
+
+        action_pred, _ = self._inverse_net(inputs=[prev_feature, feature])
+
+        if tensor_spec.is_discrete(self._action_spec):
+            inverse_loss = tf.nn.softmax_cross_entropy_with_logits(
+                labels=prev_action, logits=action_pred)
+        else:
+            inverse_loss = 0.5 * tf.reduce_mean(
+                tf.square(prev_action - action_pred), axis=-1)
+
+        intrinsic_reward = tf.stop_gradient(forward_loss)
+        intrinsic_reward = self._reward_normalizer.normalize(intrinsic_reward)
```

Comment:
with this normalization, the intrinsic reward could be negative?? likely it could mess up with lots of other stuff.

Reply:
Guaranteed positive reward could mess up with other stuff too. For example, for mountain car task, the original goal is to reach the goal as soon as possible. With a positive goal, the agent has less incentive to terminate the episode by reaching the goal.